### PR TITLE
Reducer/Saga changes to support updating the notes for a collection add-on

### PR DESCRIPTION
**Pull request by bobsilverberg**
Ref: #4578 
~~Depends on: #5165~~

### CODE REVIEW

**Comment by bobsilverberg** [comment]
@willdurand After spending some time on the UI portion of this feature, I think I want to make some changes to this approach. The only thing we ever do when updating a collection add-on is to add/change/remove notes, so I'm thinking that the reducer actions should be named thusly. I.e., `updateCollectionAddonNote`, as opposed to just `updateCollectionAddon`, and I also think it's worth adding a reducer method for `removeCollectionAddonNote`. As it stands, the implementation for `remove` is to just perform an `update` with a `notes` value of `''`, but I think it makes sense to encapsulate the logic for that in the reducer with its own method.

What do you think?

**Comment by willdurand** [comment]
> @willdurand After spending some time on the UI portion of this feature, I think I want to make some changes to this approach. The only thing we ever do when updating a collection add-on is to add/change/remove notes, so I'm thinking that the reducer actions should be named thusly. I.e., updateCollectionAddonNote, as opposed to just updateCollectionAddon

@bobsilverberg that makes a lot of sense.

> and I also think it's worth adding a reducer method for removeCollectionAddonNote. As it stands, the implementation for remove is to just perform an update with a notes value of '', but I think it makes sense to encapsulate the logic for that in the reducer with its own method.

well, this is up to you. Usually, we extract the logic of the switch/case when it is too big (like > 15-20 LOC) and/or reused for another action.

**Comment by bobsilverberg** [comment]
> well, this is up to you. Usually, we extract the logic of the switch/case when it is too big (like > 15-20 LOC) and/or reused for another action.

My thought was more about encapsulating the logic for removing a note in the reducer, rather than having it in the calling component. If we just have `updateCollectionAddonNote` then the caller would need to know that, in order to actually remove a note, they need to pass a `''` as the value of `notes`, and it seems like it would be better for a caller to not have to know that. It also opens the door later to a different implementation of removing notes if that changes (which it may).

**Comment by willdurand** [comment]
> My thought was more about encapsulating the logic for removing a note in the reducer, rather than having it in the calling component. If we just have updateCollectionAddonNote then the caller would need to know that, in order to actually remove a note, they need to pass a '' as the value of notes, and it seems like it would be better for a caller to not have to know that. It also opens the door later to a different implementation of removing notes if that changes (which it may).

Ah. I see, sorry. Having two different actions is a better design indeed.

**Comment by bobsilverberg** [comment]
I changed my mind and am not going to rename the method, because all I'd be doing is making the name longer, but I did introduce a separate reducer method for deleting the notes for a collection add-on, which simply makes use of the existing saga for updating a collection add-on. This has also been rebased against the changes from #5165, so should be ready for a review.

**Comment by willdurand** [comment]
> I did introduce a separate reducer method for deleting the notes for a collection add-on, which simply makes use of the existing saga for updating a collection add-on

I would prefer two different actions and action creators (what you call "reducer method" I guess) bound to the same saga because it is a bit strange to have an `UPDATE` that is actually a `DELETE` in the redux logs.

The API documentation indicates that `null` can be sent, and I wonder if that would make sense to send `null` rather than an empty string.

**Comment by bobsilverberg** [comment]
> I would prefer two different actions and action creators (what you call "reducer method" I guess) bound to the same saga because it is a bit strange to have an UPDATE that is actually a DELETE in the redux logs.

Sure, that makes sense.

> The API documentation indicates that null can be sent, and I wonder if that would make sense to send null rather than an empty string.

We cannot because of https://github.com/mozilla/addons-server/issues/7832. Using `''` is a workaround for now. I will add a comment to the code pointing to that issue.

**Comment by willdurand** [comment]
> We cannot because of mozilla/addons-server#7832. Using '' is a workaround for now. I will add a comment to the code pointing to that issue.

ah sorry, thanks for letting me know!

**Comment by bobsilverberg** [comment]
@willdurand I have made the changes we discussed, and have also added the ability for the collections reducer to load `collectionNotes` for an add-on that is in a collection, as we need those both for displaying the notes and also for creating/editing/removing the notes. This is ready for another look.

File: src/amo/reducers/collections.js [hunk 1]
```diff
@@ -642,12 +646,77 @@ export const deleteCollection = ({
...
+    payload: {
+      addonId, errorHandlerId, notes, page, slug, username,
+    },
+  };
+};
+
+// The DeleteCollectionAddonNotesAction is identical to the
+// UpdateCollectionAddonAction, with the only difference being that
+// we pass '' into the notes field. Therefore we can reuse the
+// UpdateCollectionAddonParams type to describe the payload.
```
Comment (willdurand) [inline src/amo/reducers/collections.js]: Thanks!

File: src/amo/reducers/collections.js [hunk 2]
```diff
@@ -642,12 +646,77 @@ export const deleteCollection = ({
...
+  addonId, errorHandlerId, page, slug, username,
+}: UpdateCollectionAddonParams = {}): DeleteCollectionAddonNotesAction => {
+  invariant(addonId, 'The addonId parameter is required');
+  invariant(errorHandlerId, 'The errorHandlerId parameter is required');
+  invariant(page, 'The page parameter is required');
+  invariant(slug, 'The slug parameter is required');
+  invariant(username, 'The username parameter is required');
+
+  // For delete we set the notes field to an empty string and call update.
+  // TODO: Use a null instead when https://github.com/mozilla/addons-server/issues/7832 is fixed.
```
Comment (willdurand) [inline src/amo/reducers/collections.js]: 👍

File: src/core/types/addons.js [hunk 3]
```diff
@@ -173,6 +173,8 @@ export type AddonType = {|
   isRestartRequired: boolean,
   isWebExtension: boolean,
   themeData?: ThemeData,
+  // This is present for collection add-ons
+  collectionNotes?: string | null,
```
Comment (willdurand) [inline src/core/types/addons.js]: I don't think we should be doing this, because it weakens our `AddonType` IMO, meaning the Flow checker will be less accurate and potentially miss issues.

Since this is only used in the "collections" context, that would probably make sense to have a new type:

```js
type CollectionAddonType = {|
  ...AddonType,
  notes: string | null,
|};
```

What do you think?

Comment (willdurand) [reply src/core/types/addons.js]: I also think it would make sense to do it here.

File: tests/unit/amo/helpers.js [hunk 4]
```diff
@@ -353,15 +353,25 @@ export const createFakeCollectionDetail = ({
...
+
+export const createFakeCollectionAddons = ({
+  addons = [createFakeCollectionAddon()],
+} = {}) => {
   return addons.map((addon) => ({
-    addon,
+    addon: addon.addon,
     downloads: 0,
-    notes: null,
+    notes: addon.notes,
```
Comment (willdurand) [inline tests/unit/amo/helpers.js]: You could also destructure the unique `map` parameter: `.map({ addon, notes }) => ({`

File: tests/unit/amo/reducers/test_collections.js [hunk 5]
```diff
@@ -169,8 +171,10 @@ describe(__filename, () => {
         detail: collectionDetail,
       }));
 
+      const fakeCollectionAddon =
+        createFakeCollectionAddon({ addon: { ...fakeAddon, id: 333 }, notes });
```
Comment (willdurand) [inline tests/unit/amo/reducers/test_collections.js]: I usually find more readable to write:

```js
const fakeCollectionAddon = createFakeCollectionAddon({
  addon: { ...fakeAddon, id: 333 },
  notes,
});
```

By having one parameter per line, I find it easier to parse the different parameters at a given level (like here `addon` and `note` are the two params given to `createFakeCollectionAddon`).

File: tests/unit/amo/reducers/test_collections.js [hunk 6]
```diff
@@ -444,6 +451,24 @@ describe(__filename, () => {
       expect(collectionInState.addons)
         .toEqual(createInternalAddons(addons));
     });
+
+    it('loads notes for collection add-ons', () => {
+      const notes = 'These are some notes.';
+      const fakeCollectionAddon =
+        createFakeCollectionAddon({ notes });
```
Comment (willdurand) [inline tests/unit/amo/reducers/test_collections.js]: These two lines should fit on a single line, no?

File: tests/unit/amo/reducers/test_collections.js [hunk 7]
```diff
@@ -444,6 +451,24 @@ describe(__filename, () => {
...
+      });
+      const collection = createFakeCollectionDetail({ addons });
+
+      const state = loadCollectionIntoState({
+        state: initialState, collection, addons,
+      });
+
+      const collectionInState = state.byId[collection.id];
+      expect(collectionInState.addons[0].collectionNotes)
+        .toEqual(notes);
```
Comment (willdurand) [inline tests/unit/amo/reducers/test_collections.js]: In general, it is better to use the full available width. 80 chars is not a lot, but these two lines should have less chars than that 🤓

File: src/amo/reducers/collections.js [hunk 8]
```diff
@@ -642,12 +646,77 @@ export const deleteCollection = ({
...
+  return {
+    type: DELETE_COLLECTION_ADDON_NOTES,
+    payload: {
+      addonId, errorHandlerId, notes: '', page, slug, username,
```
Comment (willdurand) [inline src/amo/reducers/collections.js]: Would it make sense to add a test case for this action creator that returns an empty string for `notes`? (so that we notice it when we'll change to `null` once the API is fixed).

File: src/core/types/addons.js (hunk 3 above)
Comment (bobsilverberg) [reply src/core/types/addons.js]: I originally tried to create a new type for `CollectionAddonType`, but it was causing Flow nightmares. I agree it would be better, so I'll try again. 👍

File: src/amo/reducers/collections.js (hunk 8 above)
Comment (bobsilverberg) [reply src/amo/reducers/collections.js]: That is being tested here: https://github.com/bobsilverberg/addons-frontend/blob/6977e2f7448ce56d8a3640be13925b5d1ae9e2de/tests/unit/amo/sagas/test_collections.js#L880. This is generally (I believe) the way we test action creators.

Do you disagree?

Comment (willdurand) [reply src/amo/reducers/collections.js]: Yup 👍

**Comment by bobsilverberg** [comment]
Flow is happy so this is ready for another review.

File: tests/unit/amo/reducers/test_collections.js (hunk 6 above)
Comment (willdurand) [reply tests/unit/amo/reducers/test_collections.js]: This has not been changed.

Comment (bobsilverberg) [reply tests/unit/amo/reducers/test_collections.js]: I thought I changed that. Odd. I'll fix it.
